core/executor/src/wasm_executor.rs

```diff
@@ -124,1201 +140,1135 @@ fn deadline_to_timestamp(deadline: u64) -> Option<offchain::Timestamp> {
+	{
+		let args = args.as_ref().iter().copied().map(Into::into);
+		if index < Self::function_count() {
+			return self.execute_function(index, args)
+				.map_err(Error::FunctionExecution)
+				.map_err(wasmi::Trap::from)
+				.map(|v| v.map(Into::into))
+		} else if index < SubstrateExternals::function_count() + Self::function_count() {
+			return SubstrateExternals::execute_function(index - Self::function_count(), args, self)
+				.map_err(Error::FunctionExecution)
+				.map_err(wasmi::Trap::from)
+				.map(|v| v.map(Into::into))
 		} else {
-			debug_trace!(
-				target: "wasm-trace",
-				"*** Setting storage:  {} -> {}   [k={}]",
-				primitives::hexdisplay::ascii_format(&key),
-				HexDisplay::from(&value),
-				HexDisplay::from(&key),
-			);
+			Err(
+				wasmi::Trap::from(
+					Error::from(format!("Could not find host function with index: {}", index))
+				)
+			)
 		}
-		this.ext.set_storage(key, value);
-		Ok(())
-	},
-	ext_set_child_storage(
-		storage_key_data: *const u8,
-		storage_key_len: u32,
-		key_data: *const u8,
-		key_len: u32,
-		value_data: *const u8,
-		value_len: u32
-	) => {
-		let storage_key = this.memory.get(storage_key_data, storage_key_len as usize)
-			.map_err(|_| "Invalid attempt to determine storage_key in ext_set_child_storage")?;
-		let key = this.memory.get(key_data, key_len as usize)
-			.map_err(|_| "Invalid attempt to determine key in ext_set_child_storage")?;
-		let value = this.memory.get(value_data, value_len as usize)
-			.map_err(|_| "Invalid attempt to determine value in ext_set_child_storage")?;
-		if let Some(_preimage) = this.hash_lookup.get(&key) {
-			debug_trace!(
-				target: "wasm-trace", "*** Setting child storage: {} -> %{} -> {}   [k={}]",
-				primitives::hexdisplay::ascii_format(&storage_key),
-				primitives::hexdisplay::ascii_format(&_preimage),
-				HexDisplay::from(&value),
-				HexDisplay::from(&key)
-			);
-		} else {
-			debug_trace!(
-				target: "wasm-trace", "*** Setting child storage: {} ->  {} -> {}   [k={}]",
-				primitives::hexdisplay::ascii_format(&storage_key),
-				primitives::hexdisplay::ascii_format(&key),
-				HexDisplay::from(&value),
-				HexDisplay::from(&key)
-			);
+	}
+}
+
+impl_wasm_host_interface! {
+	#[inherent_externals]
+	impl FunctionExecutor where this {
+		ext_malloc(size: WordSize) -> Pointer<u8> {
+			let r = this.heap.allocate(size).map_err(|e| format!("{:?}", e))?;
+			debug_trace!(target: "sr-io", "malloc {} bytes at {}", size, r);
+			Ok(r)
+		}
+
+		ext_free(addr: Pointer<u8>) {
+			this.heap.deallocate(addr).map_err(|e| format!("{:?}", e))?;
+			debug_trace!(target: "sr-io", "free {}", addr);
+			Ok(())
+		}
+
+		ext_sandbox_instantiate(
+			dispatch_thunk_idx: u32,
+			wasm_ptr: Pointer<u8>,
+			wasm_len: WordSize,
+			imports_ptr: Pointer<u8>,
+			imports_len: WordSize,
+			state: u32,
+		) -> u32 {
+			let wasm = this.read_memory(wasm_ptr, wasm_len)
+				.map_err(|_| "OOB while ext_sandbox_instantiate: wasm")?;
```

Comment:
Could we add the function name to the error message inside the macro? Repeating it now is really tedious. Maybe an idea for the `proc-macro` you've mentioned earlier.

Reply:
Also can the functions just return `Result<_, impl Debug>`? To simplify error handling?

Reply:
Yeah, when all this stuff is auto generated, I will also attach the function names to the error messages automatically.